src/tracked_struct.rs

```diff
@@ -148,15 +148,34 @@ where
 /// struct and later moved to the [`Memo`](`crate::function::memo::Memo`).
 #[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Copy, Clone)]
 pub(crate) struct KeyStruct {
-    /// The hash of the `#[id]` fields of this struct.
-    /// Note that multiple structs may share the same hash.
-    data_hash: u64,
+    /// Tracked struct key
+    disambiguation_key: DisambiguationKey,
 
     /// The unique disambiguator assigned within the active query
-    /// to distinguish distinct tracked structs with the same hash.
+    /// to distinguish distinct tracked structs with the same key.
     disambiguator: Disambiguator,
 }
 
+impl KeyStruct {
+    pub(crate) fn ingredient_index(&self) -> IngredientIndex {
+        self.disambiguation_key.ingredient_index
+    }
+}
+
+/// Stores the data that (almost) uniquely identifies a tracked struct.
+/// This includes the ingredient index of that struct type plus the hash of its id fields.
+/// This is mapped to a disambiguator -- a value that starts as 0 but increments each round,
+/// allowing for multiple tracked structs with the same hash and ingredient_index
+/// created within the query to each have a unique id.
+#[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Copy, Clone)]
+pub struct DisambiguationKey {
```

Comment:
It seems to me that the naming of `DisambiguationKey` is not ideal, given e.g. `KeyStruct`. I'd like some names that make the relationship clearer. Here are some possibilities -- hopefully it's clear which is meant to be which in each case!

* `HashStruct` and `KeyStruct` (close to today)
* `IdentityHash` and `Identity`
* `IdentityHash` and `DisambiguatedIdentityHash`
* `IngredientAndHash` and `DisambiguatorIngredientAndHash` (just name it after the contents)

What do you think @puuuuh?

Reply:
IdentityHash and Identity is the best variant, imo

Reply:
Also something like IdentityHash-StructCompoundId?